File: service.js
const axios = require('axios');

const gql = async (query, variables = {}) => {
  let response = await axios.post('https://api.hashnode.com/', {
    query,
    variables,
  }, {
    headers: {
      'Content-Type': 'application/json',
    },
  });
  return response.data;
}

const USERS = [
  {
    hashnode: 'born2ngopi',
    naya: 'chandra',
  },
  {
    hashnode: 'hadihammurabi',
    naya: 'hadihammurabi',
  },
];

const getAllPosts = async () => {
  const getArticlePromises = USERS.map(async (user) => {
    const GET_USER_ARTICLES = `query {
      user(username: "${user.hashnode}") {
          publication {
              posts {
                  title
                  brief
                  slug
                  coverImage
                  dateAdded
              }
          }
      }
    }`;
    const response = await gql(GET_USER_ARTICLES);
    if (response.data != null
      && response.data.user != null
      && response.data.user != null
      && response.data.user.publication != null
      && response.data.user.publication.posts != null
      && response.data.user.publication.posts.length > 0
    ) {
      response.data.user.publication.posts = response.data.user.publication.posts.map((post) => {
        return {
          ...post,
          user,
        };
      });
    }
    return response;
  });
  const responses = await Promise.all(getArticlePromises);
  let posts = [];
  responses.forEach((r) => {
    if (r.data != null
      && r.data.user != null
      && r.data.user.publication != null
      && r.data.user.publication.posts != null
      && r.data.user.publication.posts.length > 0
    ) {
      r.data.user.publication.posts.forEach(post => {
        posts.push(post);
      });
    }
  });
  posts = posts.sort(function (a, b) {
    let aDate = new Date(a.dateAdded);
    let bDate = new Date(b.dateAdded);
    if (aDate < bDate) return 1;
    if (aDate > bDate) return -1;
    return 0;
  });
  return posts;
};

module.exports = {
  getAllPosts,
};

File: service.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strings"
	"sync"
)

var Users = []map[string]string{
	{"hashnode": "born2ngopi", "naya": "chandra"},
	{"hashnode": "hadihammurabi", "naya": "hadihammurabi"},
	{"hashnode": "nurkholis", "naya": "kholis"},
	{"hashnode": "hdinjos", "naya": "fullsnackdev"},
	{"hashnode": "yogahermawan", "naya": "yogahermawan"},
}

func GQL(query string) ([]byte, error) {
	cleanQuery := strings.ReplaceAll(query, "\n", "")
	cleanQuery = strings.ReplaceAll(cleanQuery, "\t", "")
	cleanQuery = strings.ReplaceAll(cleanQuery, " ", "")

	reqData := fmt.Sprintf(`{"query":"%s","variables":{}}`, cleanQuery)
	reqBody := strings.NewReader(reqData)
	resp, err := http.Post("https://api.hashnode.com/", "application/json", reqBody)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	resBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return resBody, nil
}

type Post struct {
	Title      string `json:"title"`
	Brief      string `json:"brief"`
	Slug       string `json:"slug"`
	CoverImage string `json:"coverImage"`
	DateAdded  string `json:"dateAdded"`

	UsernameHashnode string
	UsernameNaya     string
}

type Publication struct {
	Posts []Post `json:"posts"`
}

type User struct {
	Publication Publication `json:"publication"`
}

type Data struct {
	User User `json:"user"`
}

type DtoGetPosts struct {
	Data Data `json:"data"`
}

func GetPostsAndMapUsername(usernameHashnode string, usernameNaya string) ([]Post, error) {
	resp, err := GQL(fmt.Sprintf(`
	query {
		user(username: \"%s\") {
			publication {
				posts {
					title,
					brief,
					slug,
					coverImage,
					dateAdded,
				},
			},
		},
	}
	`, usernameHashnode))
	if err != nil {
		return nil, err
	}

	respToDto := DtoGetPosts{}
	err = json.Unmarshal(resp, &respToDto)
	if err != nil {
		return nil, err
	}

	posts := make([]Post, 0)
	for _, p := range respToDto.Data.User.Publication.Posts {
		p.UsernameHashnode = usernameHashnode
		p.UsernameNaya = usernameNaya
		posts = append(posts, p)
	}

	return posts, nil
}

func SortPostsByDate(posts []Post) []Post {
	sort.SliceStable(posts, func(i, j int) bool {
		return posts[i].DateAdded > posts[j].DateAdded
	})

	return posts
}

func GetAllUsersPosts() ([]Post, error) {
	posts := make([]Post, 0)
	var wg sync.WaitGroup
	for _, u := range Users {
		wg.Add(1)
		go worderGetAllUsersPosts(u["hashnode"], u["naya"], &posts, &wg)
	}

	wg.Wait()
	return SortPostsByDate(posts), nil
}

func worderGetAllUsersPosts(usernameHashnode, usernameNaya string, posts *[]Post, wg *sync.WaitGroup) {
	defer wg.Done()
	postsPerUser, err := GetPostsAndMapUsername(usernameHashnode, usernameNaya)
	if err != nil {
		return
	}
	*posts = append(*posts, postsPerUser...)
}
